src/main/java/jenkins/plugins/git/GitToolChooser.java

```diff
@@ -0,0 +1,195 @@
+package jenkins.plugins.git;
+
+import hudson.EnvVars;
+import hudson.ExtensionList;
+import hudson.ExtensionPoint;
+import hudson.Util;
+import hudson.model.Item;
+import hudson.model.TaskListener;
+import hudson.plugins.git.GitTool;
+import hudson.plugins.git.util.GitUtils;
+import jenkins.model.Jenkins;
+import org.apache.commons.io.FileUtils;
+import org.jenkinsci.plugins.gitclient.Git;
+import org.jenkinsci.plugins.gitclient.GitClient;
+import org.jenkinsci.plugins.gitclient.JGitApacheTool;
+import org.jenkinsci.plugins.gitclient.JGitTool;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+/**
+ * A class which allows Git Plugin to choose a git implementation by estimating the size of a repository from a distance
+ * without requiring a local checkout.
+ */
+public class GitToolChooser {
+
+    private long sizeOfRepo = 0L;
+    private String implementation;
+    private String gitTool;
```

Comment:
Is this decoupled from `implementation`? Should it be?

Reply:
Should this perhaps actually be a `GitTool` type? Still looking through but that seems to have the information you need.

Reply:
Great question, I didn't include GitTool because of the fear of adding more complexity, keeping a path of git executable as a string seems to work fine so far but obviously we have some issues with it and I need to test using GitTool class as an option.